### PR TITLE
Compile the model for all supported seqlen before prefill tracing

### PR DESCRIPTION
**Pull request by nostojicTT**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/31437)

### Problem description
The problem is explained in the issue linked above.

### What's changed
- Compile and record all traces that are supported at the beginning
- If we use DP, we need to capture the trace for each submesh independently
- tt_transformers does not have batched prefill, so that's why we use `batch_size = 1` in the `warmup_prefill_traces`  --when we have `batch_size = 32` in `prefill_forward_text` we would just take one prompt from the batch and process it. That's why we use `batch_size = 1` in the newly added function since the workflow is the same, unlike how it's done for llama_70b for TG where they have implemented the batched prefill, so the workflow won't be the same for `batch_size = 1` and `batch_size = 32` : https://github.com/tenstorrent/tt-metal/blob/6ff23d72ee88fc9e2fc76f80394c19d2a4abbbbd/models/demos/llama3_70b_galaxy/tt/generator.py#L89

### Checklist
- [x] [All post commit] https://github.com/tenstorrent/tt-metal/actions/runs/19235874322
- [x] (For models and ops writers) [Single-card demo tests] https://github.com/tenstorrent/tt-metal/actions/runs/19235883397
- [x] [Galaxy demo tests, for Llama] https://github.com/tenstorrent/tt-metal/actions/runs/19241011252
- [x] (For runtime and ops writers) [T3000 unit tests] https://github.com/tenstorrent/tt-metal/actions/runs/19235896397
- [x] (For models and ops writers) [T3000 demo tests] https://github.com/tenstorrent/tt-metal/actions/runs/19235902822

## Note : Tests for Llama3-70b failed in the linked tests because of insufficient trace region size, so I will link two runs where I hardcoded the `trace_region_size` for Llama-8B and Llama-70B (other models will use the default region size passed through `@parametrize` for device_params) in order to tests to pass. When my next PR gets merged (https://github.com/tenstorrent/tt-metal/pull/32086), the `trace_region_size` won't be hardcoded in the `simple_text_demo.py` when parametrizing `device_params`, instead it will be overridden per model and per arch.

Galaxy run : https://github.com/tenstorrent/tt-metal/actions/runs/19276400784
T3K demo test run : https://github.com/tenstorrent/tt-metal/actions/runs/19276406689

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -189,7 +227,7 @@ def prefill_forward_text(
 
         out_list = []
         for idx, user_id in enumerate(empty_slots):
-            model_id = user_id // max_batch_size_per_model
+            model_id = user_id // max_batch_size_per_model if not self.active_prefill_warmup else self.active_model_id
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py]: so we will execute the same `model_id` 32 times in batch=32?
is that absolutely necessary?

File: models/tt_transformers/tt/model_config.py [hunk 2]
```diff
@@ -1345,6 +1345,24 @@ def _get_xattn_kv_prefill_mem_cfg(seq_len):
             )
             logger.info(f"LM head grid: {self.lm_head_core_grid}")
 
+        self.supported_seqlens_dictionary = {
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/model_config.py]: this cannot scale. imagine 30 models and 10 platforms (wormholes, blackholes..)

Comment (ppetrovicTT) [reply models/tt_transformers/tt/model_config.py]: also, extract this into a function, so that it's easier to maintain.

**Review by ppetrovicTT — CHANGES_REQUESTED** [review]
let's try to improve the architecture / responsibilities a bit.

`prefill_forward_text` will rely on a flag to warm up, then `warmup_prefill_traces` will set other flags and call recursively `prefill_forward_text`, maybe overriding the `model_id` within through a member variable.

i think this would be very hard to maintain and debug.

fyi, vLLM v1 has a special function for model warm-up, so we will be extracting that logic from `prefill_forward_text` soon, hopefully.

therefore, i think it would be better to keep the change as encapsulated as possible. i think it's better to extend the `prefill_forward_text` signature with forced model_id.

File: models/tt_transformers/tt/generator.py (hunk 1 above)
Comment (nostojicTT) [reply models/tt_transformers/tt/generator.py]: This will only be executed once since we when we call `prefill_forward_text` from `warmup_prefill_traces` we assign `empty_slots` to be `warmup_empty_slots` which will be `list(range(1))`, so that means that we will execute the code below once only for each sequence length that is supported for each model (for each submesh created by DP)

File: models/tt_transformers/tt/model_config.py (hunk 2 above)
Comment (nostojicTT) [reply models/tt_transformers/tt/model_config.py]: This is a problem that I don't know how to tackle, because for each model on each platform we need to decide whether trace will or will not improve TTFT. As discussed with @rjakovljevicTT, we can set the supported sequence length to be up to 8K for each model even if there are no TTFT improvements just to avoid the issue with scaling. But if we do it that way, we have a new issue, because on some platforms, we can't record traces up to 8K because of space limitation (N150 example here).
So a solution to this could be:
1. (Scaling will remain a problem)
Cons:We would for each model need to enter these sequence lengths manually for each platform which would take a lot of time (measuring op-to-op gaps on every platform, measuring the needed trace region size)
Pros: We take up minimal space for trace region and also we are sure that for non-included sequence lengths there is no TTFT improvement
2. (Scaling is not a problem)
Cons: We take up more space than needed
Pros: More scalable

Comment (nostojicTT) [reply models/tt_transformers/tt/model_config.py]: What if some models take up less trace region space and we can cover the 2K ISL on N150?
Example: Qwen3-32B can store traces for ISL up to 2K tokens, but Llama-8B only up to 1K tokens? If we want to have scalable code does that mean the following:
"N150" : [128, 256, 512, 1024]
This means that for Qwen, we won't have TTFT improvements for 2K ISL even tho we could

Comment (ppetrovicTT) [reply models/tt_transformers/tt/model_config.py]: From an offline conversation: Let's have a default and we can override where needed.

**Review by gwangTT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -62,6 +62,34 @@ def __init__(self, model, model_args, mesh_device, processor=None, tokenizer=Non
...
+
+    def warmup_prefill_traces(
+        self,
+        page_table,
+        kv_cache,
+        enable_trace,
+    ):
+        self.prefill_traces_warmup = True
+        for model_id in range(self.data_parallel):
+            for supported_length in [128, 256, 512, 1024, 2048, 4096, 8192]:
```
Comment (mtairum) [inline models/tt_transformers/tt/generator.py:95]: Can we created a `trace_prefill_lengths` dictionary in `model_config.py` instead? So each model+arch can have it's own list?

Supported length is a misnomer, since all lengths are supported, these are just all the ones that fit together due to available DRAM. This is going to vary between each model + arch. If we do the same for every model this is going to break rapidly.

File: models/tt_transformers/tt/generator.py [hunk 4]
```diff
@@ -62,6 +62,34 @@ def __init__(self, model, model_args, mesh_device, processor=None, tokenizer=Non
...
+                # Only sequence lengths used by Llama-3.1-8B since we only support trace for Llama-3.1-8B for now
+                warmup_tokens = torch.zeros(1, supported_length, dtype=torch.long)
+                warmup_prompt_lens = torch.tensor([supported_length], dtype=torch.long)
+                warmup_empty_slots = list(range(1))
+                if not self.model_args[0].can_enable_trace(supported_length):
+                    continue
```
Comment (mtairum) [inline models/tt_transformers/tt/generator.py:103]: Add a TODO comment here saying we're working on adding trace support for all models shortly.

File: models/tt_transformers/tt/generator.py [hunk 5]
```diff
@@ -177,6 +206,13 @@ def prefill_forward_text(
             # Only paged attention is supported for prefill
             enable_trace = False
 
+        if self.prefill_traces_warmup is False and enable_trace:
```
[the remark this answers is not in the document]
Comment (mtairum) [reply models/tt_transformers/tt/generator.py]: Good suggestion.

File: models/tt_transformers/tt/model_config.py [hunk 6]
```diff
@@ -1846,7 +1846,11 @@ def can_enable_trace(self, prefill_seq_len):
         else:
             allowed_seq_lens = [128, 256, 512, 1024, 2048, 4096, 8192]
```
Comment (mtairum) [inline models/tt_transformers/tt/model_config.py:1848]: I know this is not part of the PR, but regarding the dict change I mention, let's address that with this as well.

**Review by mtairum — APPROVED** [review]
Pre-approving to unblock but please either address the comments now or create a follow-up issue to fix them.

File: models/tt_transformers/tt/generator.py (hunk 3 above)
Comment (nostojicTT) [reply models/tt_transformers/tt/generator.py:95]: Will create that in the next PR and add it to `model_config.py`

**Review by yieldthought — APPROVED** [review]
Subject to handling miguel's comments, follow-on PR is ok if it happens right away

File: models/tt_transformers/tt/generator.py [hunk 7]
```diff
@@ -79,6 +79,36 @@ def __init__(self, model, model_args, mesh_device, processor=None, tokenizer=Non
...
+    def warmup_prefill_traces(
+        self,
+        page_table,
+        kv_cache,
+        enable_trace,
+    ):
+        self.prefill_traces_warmup = True
+        for model_id in range(self.data_parallel):
+            for supported_length in [128, 256, 512, 1024, 2048, 4096, 8192]:
+                # Only sequence lengths used by Llama-3.1-8B since we only support trace for Llama-3.1-8B for now
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py:96]: Llama-3.3-70B was enabled a few days ago: https://github.com/tenstorrent/tt-metal/commit/e47fe9a3417752ed992e315388f2cf8c1903ad2b.
This is just to check if you've covered that model as well.

File: models/tt_transformers/tt/generator.py [hunk 8]
```diff
@@ -195,6 +226,13 @@ def prefill_forward_text(
             # Only paged attention is supported for prefill
             enable_trace = False
 
+        if not self.prefill_traces_warmup and enable_trace:
+            self.warmup_prefill_traces(
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py]: nitpick: if you named the function `try_warmup_prefill_traces` you could check `self.prefill_traces_warmup` inside and everything about warming up would be self contained. (it also checks `enable_trace` inside)

**Review by ppetrovicTT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py (hunk 7 above)
Comment (nostojicTT) [reply models/tt_transformers/tt/generator.py:96]: These sequence lengths are also used in Llama-3.3-70B so I can leave it like this until I merge the dictionary PR.
